#### Import and explore the data


In [222]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [223]:
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


import numpy as np
import seaborn as sns
import statsmodels.api as sm
import matplotlib.pyplot as plt
#%matplotlib inline
sns.set(style="whitegrid")

from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator, TransformerMixin



import warnings 
warnings.filterwarnings('ignore')

##### Custome Class

In [224]:
# class Columns(BaseEstimator, TransformerMixin):
#     def __init__(self, columns):
#         self.columns = columns

#     def fit(self, X, y=None):
#         return self

#     def transform(self, X, y=None):
#         return X[self.columns]

def check_distribution(traindata,testdata,columns_list):
    for column_name in columns_list:
        print("---------------",column_name,"------------")
        print(round(traindata[column_name].value_counts() / traindata[column_name].count(),2)*100)
        print(round(testdata[column_name].value_counts() / testdata[column_name].count(),2)*100)

##### Reading Data

In [225]:
path = "/content/drive/MyDrive/Colab Notebooks/Greatelearning/"
#path = "/Data/"
data = pd.read_csv(path+"signal-data.csv")
data.head()

Time        0        1          2          3       4      5  \
0  2008-07-19 11:55:00  3030.93  2564.00  2187.7333  1411.1265  1.3602  100.0   
1  2008-07-19 12:32:00  3095.78  2465.14  2230.4222  1463.6606  0.8294  100.0   
2  2008-07-19 13:17:00  2932.61  2559.94  2186.4111  1698.0172  1.5102  100.0   
3  2008-07-19 14:43:00  2988.72  2479.90  2199.0333   909.7926  1.3204  100.0   
4  2008-07-19 15:22:00  3032.24  2502.87  2233.3667  1326.5200  1.5334  100.0   

          6       7       8       9      10      11        12   13       14  \
0   97.6133  0.1242  1.5005  0.0162 -0.0034  0.9455  202.4396  0.0   7.9558   
1  102.3433  0.1247  1.4966 -0.0005 -0.0148  0.9627  200.5470  0.0  10.1548   
2   95.4878  0.1241  1.4436  0.0041  0.0013  0.9615  202.0179  0.0   9.5157   
3  104.2367  0.1217  1.4882 -0.0124 -0.0033  0.9629  201.8482  0.0   9.6052   
4  100.3967  0.1235  1.5031 -0.0031 -0.0072  0.9569  201.9424  0.0  10.5661   

         15       16      17        18       19      20       21       22  \
0  414.8710  10.0433  0.9680  192.3963  12.5190  1.4026 -5419.00  2916.50   
1  414.7347   9.2599  0.9701  191.2872  12.4608  1.3825 -5441.50  2604.25   
2  416.7075   9.3144  0.9674  192.7035  12.5404  1.4123 -5447.75  2701.75   
3  422.2894   9.6924  0.9687  192.1557  12.4782  1.4011 -5468.25  2648.25   
4  420.5925  10.3387  0.9735  191.6037  12.4735  1.3888 -5476.25  2635.25   

        23       24      25      26      27       28      29      30      31  \
0 -4043.75   751.00  0.8955  1.7730  3.0490  64.2333  2.0222  0.1632  3.5191   
1 -3498.75 -1640.25  1.2973  2.0143  7.3900  68.4222  2.2667  0.2102  3.4171   
2 -4047.00 -1916.50  1.3122  2.0295  7.5788  67.1333  2.3333  0.1734  3.5986   
3 -4515.00 -1657.25  1.3137  2.0038  7.3145  62.9333  2.6444  0.2071  3.3813   
4 -3987.50   117.00  1.2887  1.9912  7.2748  62.8333  3.1556  0.2696  3.2728   

        32      33       34       35       36       37       38        39  \
0  83.3971  9.5126  50.6170  64.2588  49.3830  66.3141  86.9555  117.5132   
1  84.9052  9.7997  50.6596  64.2828  49.3404  64.9193  87.5241  118.1188   
2  84.7569  8.6590  50.1530  64.1114  49.8470  65.8389  84.7327  118.6128   
3  84.9105  8.6789  50.5100  64.1125  49.4900  65.1951  86.6867  117.0442   
4  86.3269  8.7677  50.2480  64.1511  49.7520  66.1542  86.1468  121.4364   

      40     41    42        43       44        45        46      47  \
0  61.29  4.515  70.0  352.7173  10.1841  130.3691  723.3092  1.3072   
1  78.25  2.773  70.0  352.2445  10.0373  133.1727  724.8264  1.2887   
2  14.37  5.434  70.0  364.3782   9.8783  131.8027  734.7924  1.2992   
3  76.90  1.279  70.0  363.0273   9.9305  131.8027  733.8778  1.3027   
4  76.39  2.209  70.0  353.3400  10.4091  176.3136  789.7523  1.0341   

         48   49        50        51   52     53     54      55      56  \
0  141.2282  1.0  624.3145  218.3174  0.0  4.592  4.841  2834.0  0.9317   
1  145.8445  1.0  631.2618  205.1695  0.0  4.590  4.842  2853.0  0.9324   
2  141.0845  1.0  637.2655  185.7574  0.0  4.486  4.748  2936.0  0.9139   
3  142.5427  1.0  637.3727  189.9079  0.0  4.486  4.748  2936.0  0.9139   
4  138.0882  1.0  667.7418  233.5491  0.0  4.624  4.894  2865.0  0.9298   

       57      58       59        60       61        62       63       64  \
0  0.9484  4.7057  -1.7264  350.9264  10.6231  108.6427  16.1445  21.7264   
1  0.9479  4.6820   0.8073  352.0073  10.3092  113.9800  10.9036  19.1927   
2  0.9447  4.5873  23.8245  364.5364  10.1685  115.6273  11.3019  16.1755   
3  0.9447  4.5873  24.3791  361.4582  10.2112  116.1818  13.5597  15.6209   
4  0.9449  4.6414 -12.2945  355.0809   9.7948  144.0191  21.9782  32.2945   

        65        66      67        68   69        70       71        72  \
0  29.5367  693.7724  0.9226  148.6009  1.0  608.1700  84.0793       NaN   
1  27.6301  697.1964  1.1598  154.3709  1.0  620.3582  82.3494       NaN   
2  24.2829  710.5095  0.8694  145.8000  1.0  625.9636  84.7681  140.6972  

##### Exploratory Data Analysis



###### shape of the data

In [226]:
data.shape

(1567, 592)

Observation
- there are 592 features and 1500+ observations

###### rename column name

In [227]:
data_analysis = data.copy()
data_analysis = data_analysis.add_prefix("dimension_")
data_analysis.rename(columns={"dimension_Pass/Fail": "Target"},inplace = True)
data_analysis.columns

Index(['dimension_Time', 'dimension_0', 'dimension_1', 'dimension_2',
       'dimension_3', 'dimension_4', 'dimension_5', 'dimension_6',
       'dimension_7', 'dimension_8',
       ...
       'dimension_581', 'dimension_582', 'dimension_583', 'dimension_584',
       'dimension_585', 'dimension_586', 'dimension_587', 'dimension_588',
       'dimension_589', 'Target'],
      dtype='object', length=592)

###### Treating Target column

In [228]:
print(data_analysis["Target"].isna().sum()) # No missing value in target 
print(data_analysis["Target"].value_counts())

# Update Fail (1) = 0 and Pass (-1) = 1 for the better understanding
data_analysis["Target"] = np.where(data_analysis["Target"] == -1,1,0 )
print(data_analysis["Target"].value_counts())

print(round(data_analysis['Target'].value_counts() / data_analysis['Target'].count(),2)*100)

0
-1    1463
 1     104
Name: Target, dtype: int64
1    1463
0     104
Name: Target, dtype: int64
1    93.0
0     7.0
Name: Target, dtype: float64


 - Observation 
  - To avoid cognitive load , target value -1 is converted to 1 ( which represent pass) and 1 is converted to 0 ( which is fail)
  - 93% of observation represent the pass and 7 % represent fail

###### Missing data & Impute with Zero 

- Missing Values can be imputed by zero. 
  - Reason - since datas are collected from sensors , which has high propability for automated way of collecting. 
  - If the there no signal it can be empty , se we can impute as zero

In [229]:
print("# of data Features has missing Value",data_analysis.isna().sum()[data_analysis.isna().sum() >=1].count())

# of data Features has missing Value 538


In [230]:
data_analysis.fillna(0,inplace=True)
print("Missing value count -- ",train_data.isna().sum().sum())

Missing value count --  0


###### Split Train and test data

In [231]:
X=data_analysis.loc[:,data_analysis.columns !=  "Target"]
y=data_analysis.loc[:,data_analysis.columns ==  "Target"]
train_data_X,test_data_X,train_data_y,test_data_y = train_test_split(X,y,test_size=0.30, random_state=123)

In [232]:
train_data = pd.DataFrame(train_data_X,columns=data_analysis.columns)
train_data["Target"] = train_data_y
train_data.head()

dimension_Time  dimension_0  dimension_1  dimension_2  dimension_3  \
375  2008-08-21 18:53:00      3045.73      2479.93    2197.6778    1056.7817   
807  2008-12-09 13:08:00      2991.25      2519.26    2194.8000    1456.0760   
178  2008-10-08 00:09:00      2846.07      2514.66    2234.6111    1590.1699   
130  2008-06-08 03:22:00      3017.50      2634.30    2183.4556     955.9073   
708  2008-03-09 00:58:00      3005.53      2479.24    2171.3222     966.5755   

     dimension_4  dimension_5  dimension_6  dimension_7  dimension_8  \
375       1.3168        100.0     102.9611       0.1203       1.3814   
807       1.5130        100.0      99.0344       0.1233       1.4745   
178       1.2049        100.0      98.7567       0.1208       1.5622   
130       1.1048        100.0     102.6978       0.1223       1.4700   
708       0.8066        100.0     107.1700       0.1242       1.4924   

     dimension_9  dimension_10  dimension_11  dimension_12  dimension_13  \
375       0.0013       -0.0046        0.9749      196.9541           0.0   
807      -0.0053        0.0080        0.9698      198.3019           0.0   
178       0.0167       -0.0117        0.9698      198.1170           0.0   
130      -0.0092        0.0083        0.9514      210.6180           0.0   
708      -0.0185        0.0057        0.9668      205.0389           0.0   

     dimension_14  dimension_15  dimension_16  dimension_17  dimension_18  \
375       14.2942      428.5513        9.7666        0.9787      187.1876   
807        5.1992      410.8906        9.3152        0.9730      188.9867   
178       14.0406      414.3115       10.6319        0.9798      187.4852   
130        5.8718      420.8003       10.3046        0.9563      200.3134   
708       11.9662      414.8114        9.5805        0.9725      195.4583   

     dimension_19  dimension_20  dimension_21  dimension_22  dimension_23  \
375       12.4661        1.3965      -6402.75       3036.75      -4156.25   
807       12.4308        1.4080      -5611.75       2679.50      -4337.50   
178       12.5724        1.3944      -5442.25       2660.25      -3767.25   
130       12.4659        1.4119      -5531.50       2643.75      -4494.50   
708       12.5094        1.3860      -6233.50       2875.75      -4262.00   

     dimension_24  dimension_25  dimension_26  dimension_27  dimension_28  \
375     1578.0000        0.9862        1.8855        5.1365       65.5000   
807    -2694.6667        1.3030        1.9665        7.2630       75.3667   
178      486.0000        1.2800        2.0183        7.4145       68.1556   
130    -6779.0000        1.2737        2.0303        7.3520       69.3333   
708     1367.5000        0.9977        1.8585        5.0568       65.8556   

     dimension_29  dimension_30  dimension_31  dimension_32  dimension_33  \
375        2.7333        0.2079        4.6529       85.1712        8.5405   
807        2.5889        0.1867        3.3577       85.4656        8.5407   
178        2.7222        0.2164        3.5033       84.8180        8.6309   
130        2.6222        0.1774        3.5142       84.3585        8.7495   
708        1.9222        0.1472        4.6448       84.6838        9.0425   

     dimension_34  dimension_35  dimension_36  dimension_37  dimension_38  \
375       50.2488       64.1314       49.7512       66.2364       86.8378   
807       50.3651       64.2771       49.6349       65.9873       86.8223   
178       50.2512       64.2144       49.7488       65.9697       86.7969   
130       50.4362       64.2234       49.5638       65.9449       86.7325   
708       50.7493       64.4947       49.2507       67.1943       86.8006   

     dimension_39  dimension_40  dimension_41  dimension_42  dimension_43  \
375      118.4904         76.57         3.732          70.0      349.2918   
807      119.5174         80.93         3.674          70.0      349.3591   
178      117.4996         78.31         2.408          70.0      357.0282   
130      117.707

In [233]:
test_data = pd.DataFrame(test_data_X,columns=data_analysis.columns)
test_data["Target"] = test_data_y
test_data.head()

dimension_Time  dimension_0  dimension_1  dimension_2  dimension_3  \
906   2008-09-19 09:12:00      2966.39      2507.22    2171.5000    1028.4440   
987   2008-09-21 14:31:00      3079.44      2395.39    2209.0889    1459.4739   
929   2008-09-19 21:59:00      2914.86      2465.11    2210.2778    2120.5760   
1118  2008-09-27 01:29:00      3052.22      2496.49    2219.7667    2086.4710   
623   2008-01-09 05:32:00      3071.80      2517.65    2200.0666    1012.6747   

      dimension_4  dimension_5  dimension_6  dimension_7  dimension_8  \
906        0.7899        100.0     101.5122       0.1224       1.3487   
987        2.1612        100.0      98.9011       0.1242       1.5335   
929        1.0700        100.0      95.1089       0.1230       1.5817   
1118       1.3381        100.0      98.8900       0.1234       1.3088   
623        1.3954        100.0     103.0644       0.1212       1.3949   

      dimension_9  dimension_10  dimension_11  dimension_12  dimension_13  \
906        0.0050        0.0006        0.9632      201.9133           0.0   
987        0.0158       -0.0095        0.9804      200.2872           0.0   
929        0.0118        0.0000        0.9686      196.5141           0.0   
1118       0.0105       -0.0059        0.9846      197.3422           0.0   
623        0.0065       -0.0018        0.9665      201.0295           0.0   

      dimension_14  dimension_15  dimension_16  dimension_17  dimension_18  \
906         9.3365      412.8410       10.0515        0.9784      191.8618   
987         9.3252      400.9785        8.5730        0.9618      191.7142   
929        10.6901      417.2738        9.8875        0.9798      186.6265   
1118       10.9365      412.3116        9.5578        0.9814      187.7844   
623         9.1238      399.5597        9.6269        0.9733      191.4026   

      dimension_19  dimension_20  dimension_21  dimension_22  dimension_23  \
906        12.5504        1.4127      -5526.00       3397.75      -4184.50   
987        12.5331        1.4181      -6445.00       2984.75       -567.75   
929        12.5376        1.4086      -5508.75       2608.75      -2091.25   
1118       12.5786        1.4128      -5573.00       2719.00      -1739.50   
623        12.5302        1.4134      -6386.75       2901.75      -3755.75   

      dimension_24  dimension_25  dimension_26  dimension_27  dimension_28  \
906       -1684.25        1.3050        1.9962        7.2742       69.9222   
987        1295.75        0.9913        1.8847        5.1750       64.5889   
929        1526.00        1.2710        1.9930        7.1638       69.0556   
1118       -112.75        1.3117        1.9992        7.2533       69.5111   
623        1479.75        0.9950        1.8670        5.1130       67.2667   

      dimension_29  dimension_30  dimension_31  dimension_32  dimension_33  \
906         2.6556        0.2218        3.3880       85.4838        8.7401   
987         3.0000        0.1900        4.6724       86.0222        9.2016   
929         2.3778        0.1938        3.4174       84.4061        8.6772   
1118        2.5444        0.1923        3.3796       84.7646        8.5862   
623         2.5111        0.1959        4.6716       84.6838        9.0425   

      dimension_34  dimension_35  dimension_36  dimension_37  dimension_38  \
906        50.0631       64.3756       49.9369       66.1541       86.8963   
987        50.5239       64.0447       49.4761       67.0499       86.8786   
929        50.4095       64.2376       49.5905       66.2913       86.2590   
1118       50.4865       64.3404       49.5135       66.2443       86.5384   
623        50.7493       64.4947       49.2507       67.1943       86.8006   

      dimension_39  dimension_40  dimension_41  dimension_42  dimension_43  \
906       118.0474         78.53         2.821          70.0      353.4900   
987       118.1136         76.88         2.483          70.0      349.5127   
929       118.5420         79.86         2.

In [234]:
check_distribution(train_data,test_data,["Target"])  # Check the dristribution of the data ion Train and test 

--------------- Target ------------
1    93.0
0     7.0
Name: Target, dtype: float64
1    94.0
0     6.0
Name: Target, dtype: float64


###### removing irrelavent columns

In [235]:
train_data.drop(columns=["dimension_Time"],inplace=True) # removing Time attribute

In [236]:
# removing feature with same value for entire observation
data_stat = pd.DataFrame(data = train_data.describe(include="all").T.reset_index())
feature_name_with_same_value = list(data_stat[data_stat["min"] == data_stat["max"]] ["index"])
train_data.drop(columns=feature_name_with_same_value,inplace=True)

# Check is any duplicate 
print("is any duplicate records", train_data.duplicated().sum() )

print("Shape of the train_set after removal",train_data.shape)


is any duplicate records 0
Shape of the train_set after removal (1096, 477)


##### Handling Outliers

In [237]:
# Counting outlier points 
def checkoutliers(train_data):
  Q1 = train_data.quantile(0.25)
  Q3 = train_data.quantile(0.75)
  IQR = Q3 - Q1 
  outlier_data = pd.DataFrame(((train_data < (Q1 - 1.5 * IQR)) | (train_data > (Q3 + 1.5 * IQR))).sum(),columns=["outlier_count"])
  outlier_data.reset_index(inplace=True)
  outlier_data.rename(columns={"index":"feature"},inplace=True)
  outlier_data = outlier_data.sort_values(by="outlier_count")
  outlier_data =  outlier_data[outlier_data['outlier_count']>=1]
  return outlier_data
  

In [238]:
outlier_data = checkoutliers(train_data)
outlier_data.describe().T

count       mean        std  min   25%   50%    75%    max
outlier_count  450.0  46.708889  45.776567  1.0  15.0  35.0  60.75  270.0

In [ ]:
cutoff = round(train_data.shape[0]*0.05)
print("Number of features has less than 5% of outliers", 
      outlier_data[(outlier_data['outlier_count']<=cutoff) & (outlier_data['outlier_count']>=1)].count())
outlier_feature_name = list(outlier_data[(outlier_data['outlier_count']<=cutoff) & (outlier_data['outlier_count']>=1)]["feature"])
print(outlier_feature_name)
for feature_name in outlier_feature_name:
  perc = train_data[feature_name].quantile([0.01,0.99]).values
  #print(feature_name)
  #print(perc)
  train_data[feature_name] = np.clip(train_data[feature_name],perc[0],perc[1])




In [250]:
outlier_data = checkoutliers(train_data)
outlier_data


feature  outlier_count
292  dimension_345             12
50    dimension_54             12
241  dimension_283             12
184  dimension_200             12
191  dimension_207             12
260  dimension_303             12
437  dimension_551             12
82    dimension_86             12
7      dimension_7             13
160  dimension_167             13
245  dimension_288             13
173  dimension_182             13
432  dimension_546             13
267  dimension_310             14
290  dimension_343             14
146  dimension_153             14
259  dimension_302             14
80    dimension_84             14
192  dimension_208             14
464  dimension_578             14
348  dimension_426             14
161  dimension_168             15
418  dimension_521             15
368  dimension_446             15
331  dimension_407             15
126  dimension_131             15
42    dimension_44             15
98   dimension_103             16
322  dimension_387             16
219  dimension_249             16
53    dimension_57             16
109  dimension_114             16
182  dimension_198             17
9      dimension_9             17
315  dimension_376             17
207  dimension_223             17
144  dimension_151             17
117  dimension_122             17
303  dimension_356             18
316  dimension_377             18
199  dimension_215             18
19    dimension_20             18
243  dimension_286             18
427  dimension_541             18
361  dimension_439             18
116  dimension_121             18
118  dimension_123             18
213  dimension_239             19
128  dimension_133             19
454  dimension_568             19
108  dimension_113             19
49    dimension_53             19
94    dimension_99             19
382  dimension_470             19
346  dimension_424             19
228  dimension_269             19
270  dimension_316             20
212  dimension_238             20
16    dimension_17             20
31    dimension_32             20
73    dimension_77             21
84    dimension_88             21
2      dimension_2             21
122  dimension_127             21
168  dimension_175             21
281  dimension_334             21
58    dimension_62             21
323  dimension_388             22
472  dimension_586             22
450  dimension_564             22
52    dimension_56             22
392  dimension_480             22
310  dimension_363             23
406  dimension_495             23
452  dimension_566             23
298  dimension_351             23
441  dimension_555             23
440  dimension_554             23
120  dimension_125             23
103  dimension_108             23
71    dimension_75             23
91    dimension_95             24
275  dimension_321             24
202  dimension_218             24
197  dimension_213             25
293  dimension_346             25
95   dimension_100             25
10    dimension_10             25
236  dimension_278             25
220  dimension_250             26
229  dimension_270             26
99   dimension_104             26
301  dimension_354             26
247  dimension_290             26
76    dimension_80             27
110  dimension_115             27
113  dimension_118             27
419  dimension_522             27
78    dimension_82             27
408  dimension_497             27
238  dimension_280             28
398  dimension_487             28
401  dimension_490             28
473  dimension_587             28
121  dimension_126             29
284  dimension_337             29
83    dimension_87             29
373  dimension_454             30
274  dimension_320             30
123  dimension_128             30
360  dimension_438             31
15    dimension_16             31
166  dimension_173             31
359  dimension_437             31
265  dimension_308             31
176  dimension_185             31
186  dimension_202             32

In [247]:
perc = train_data["dimension_542"].quantile([0.01,0.99])
print(perc.values)
np.clip(train_data["dimension_542"],perc[0],perc[1])


[0.106585 0.118   ]


KeyError: ignored

- 450 feature has outliers 
- 322 features has less than 5% of there points as outliers. 
  - Capping 

#### Validation Set preparation

In [ ]:
X_train_test_validate = data_impute.iloc[:,:-1]
y_train_test_validate = data_impute.iloc[:,-1:]

X_train_test, X_validate,y_train_test,y_validate = train_test_split(X_train_test_validate, y_train_test_validate, test_size=0.15, random_state=123)
check_distribution(y_train_test,y_validate,['Target'])


NameError: ignored